format/udf-spec.md

```diff
@@ -0,0 +1,300 @@
+
+Notes:
+1. `sig1-<base32(SHA-256(signature))[:26]>` is a fixed-length hash used when the canonical signature is too long.
+It’s generated by taking the SHA-256 of the normalized signature, encoding it in Base32, keeping the first 26 characters,
+and prefixing with `sig1-`. This yields a 31-character deterministic ID, easy to verify and future-proof via the `sig1-`
+version prefix.
+
+### Parameter
+| Requirement | Field  | Type                                                                                                                                                           | Description              |
+|-------------|--------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|
+| *required*  | `name` | `string`                                                                                                                                                       | Parameter name.          |
+| *required*  | `type` | [JSON representation](https://iceberg.apache.org/spec/#appendix-c-json-serialization) of an Iceberg type (`string` for primitives, `object` for complex types) | Parameter data type.     |
+| *optional*  | `doc`  | `string`                                                                                                                                                       | Parameter documentation. |
+
+Notes:
+1. Function definitions are identified by the tuple of `type`s and there can be only one definition for a given tuple.
+2. Parameter `name`s are immutable since named argument invocation is supported (e.g., `foo(a => 1, b => 2)`). Only `doc` can be updated in place. 
+3. Variadic (vararg) parameters are not supported. Each definition must declare a fixed number of parameters.
+4. Each parameter input MUST be assignable to its declared Iceberg type. For complex types, the value’s
+   structure must match (correct field names, element/key/value types, and nesting). If a parameter—or any nested
+   field/element—is marked required, engines MUST reject null at that position (including inside structs, lists, and maps).
+5. The `return-type` is immutable. To change it, users must create a new definition and remove the old one.
+6. The function MUST return a value assignable to the declared `return-type`, meaning the returned value’s type and
+   structure must match the declared Iceberg type (including field names, element types, and nesting for complex types),
+   and any field or element marked as required MUST NOT be null. Engines MUST reject results that violate these rules.
+
+### Definition-Version
+
+Each definition can evolve over time by introducing new versions.  
```

Comment:
Shall we put a cross-version consistency rule in this section:

- function-type MUST NOT change across versions of the same definition.
- parameter types and names, return-type  MUST remain identical across versions of a given definition. Changing them requires a new definition.

Reply:
That's a good point. Updated this section to the intent clearer and more explicit.